### PR TITLE
Update the example

### LINKED ISSUE #415: Rewrite the example using ZIO libraries
With zio-json and zio-http out there, it makes sense to rewrite the example using them.

For the sake of simplicity, the implementer should preserve the existing scenario, i.e., retrieving project contributors via GitHub API. Once the example is ported, we can think about making additional scenarios.

### PR DESCRIPTION
**Pull request by paulpdaniels**
Closes #415

### CODE REVIEW

File: example/src/main/scala/example/ApiError.scala [hunk 1]
```diff
@@ -1,21 +1,21 @@
...
+import scala.util.control.NoStackTrace
 
 sealed trait ApiError extends NoStackTrace
 
 object ApiError {
-  case object CacheMiss         extends ApiError
-  case object CorruptedData     extends ApiError
-  case object GithubUnreachable extends ApiError
-  case object UnknownProject    extends ApiError
+  case class CacheMiss(key: String)       extends ApiError
```
Comment (mijicd) [inline example/src/main/scala/example/ApiError.scala:10]:
```suggestion
  final case class CacheMiss(key: String)       extends ApiError
```

File: example/src/main/scala/example/ApiError.scala [hunk 2]
```diff
@@ -1,21 +1,21 @@
...
+  case object CorruptedData               extends ApiError
+  case object GithubUnreachable           extends ApiError
+  case class UnknownProject(path: String) extends ApiError
```
Comment (mijicd) [inline example/src/main/scala/example/ApiError.scala:13]:
```suggestion
  final case class UnknownProject(path: String) extends ApiError
```

File: example/src/main/scala/example/Main.scala [hunk 3]
```diff
@@ -1,51 +1,52 @@
...
+  val cache = redis ++ sttp >>> ContributorsCache.live
 
-    (actorSystem ++ serverConfig ++ (api >>> routes)) >>> HttpServer.live
-  }
+  val runServer =
+    getConfig[ServerConfig]
+      .flatMap(conf =>
+        (Server.port(conf.port) ++ Api.server).make
+          .use_(putStrLn("Server online.") *> ZIO.never)
+      )
```
Comment (mijicd) [inline example/src/main/scala/example/Main.scala:40]: nit: I'd inline this in `run`.

File: example/src/main/scala/example/config/ServerConfig.scala [hunk 4]
```diff
@@ -0,0 +1,3 @@
+package example.config
+
+case class ServerConfig(port: Int)
```
Comment (mijicd) [inline example/src/main/scala/example/config/ServerConfig.scala:3]:
```suggestion
final case class ServerConfig(port: Int)
```

File: example/src/main/scala/example/Main.scala [hunk 5]
```diff
@@ -1,51 +1,52 @@
...
+    ZIO.effect(ConfigFactory.load().getConfig("example")),
+    AppConfig.descriptor
+  )
 
-    val actorSystem =
-      ZManaged
-        .make(ZIO.succeed(ActorSystem("zio-redis-example")))(as => ZIO.fromFuture(_ => as.terminate()).either)
-        .toLayer
+  val serverConfig = config.narrow(_.server)
+  val redisConfig  = config.narrow(_.redis)
```
Comment (mijicd) [inline example/src/main/scala/example/Main.scala:28]: I'd make these private.

File: example/src/main/scala/example/Main.scala [hunk 6]
```diff
@@ -1,51 +1,52 @@
...
-    val codec  = ZLayer.succeed[Codec](StringUtf8Codec)
-    val redis  = Logging.ignore ++ redisConfig ++ codec >>> RedisExecutor.live
-    val sttp   = AsyncHttpClientZioBackend.layer()
-    val cache  = redis ++ sttp >>> ContributorsCache.live
-    val api    = cache >>> Api.live
-    val routes = ZLayer.fromService[Api.Service, Route](_.routes)
+  val codec = ZLayer.succeed[Codec](StringUtf8Codec)
+  val redis = Logging.ignore ++ redisConfig ++ codec >>> RedisExecutor.live
+  val sttp  = AsyncHttpClientZioBackend.layer()
+  val cache = redis ++ sttp >>> ContributorsCache.live
```
Comment (mijicd) [inline example/src/main/scala/example/Main.scala:33]: Same here.

**Review by mijicd — COMMENTED** [review]
Amazing work! Just a few nitpicks and it's ready to go!

File: example/src/main/scala/example/api/Api.scala [hunk 7]
```diff
@@ -1,29 +1,22 @@
...
 import zio._
+import zhttp.http._
+import zhttp.service.Server
+import zio.json._
 
 object Api {
-  trait Service {
-    def routes: Route
+
+  val app: HttpApp[ContributorsCache, Nothing] = HttpApp.collectM {
```
Comment (anovakovic01) [inline example/src/main/scala/example/api/Api.scala:12]: This can be private too.

Comment (paulpdaniels) [reply example/src/main/scala/example/api/Api.scala]: Done

File: example/src/main/scala/example/Main.scala (hunk 6 above)
Comment (paulpdaniels) [reply example/src/main/scala/example/Main.scala]: Done

File: example/src/main/scala/example/Main.scala (hunk 5 above)
Comment (paulpdaniels) [reply example/src/main/scala/example/Main.scala]: Done

File: example/src/main/scala/example/config/ServerConfig.scala (hunk 4 above)
Comment (paulpdaniels) [reply example/src/main/scala/example/config/ServerConfig.scala]: Done

File: example/src/main/scala/example/Main.scala (hunk 3 above)
Comment (paulpdaniels) [reply example/src/main/scala/example/Main.scala]: Done

File: example/src/main/scala/example/ApiError.scala (hunk 2 above)
Comment (paulpdaniels) [reply example/src/main/scala/example/ApiError.scala]: Done

File: example/src/main/scala/example/ApiError.scala (hunk 1 above)
Comment (paulpdaniels) [reply example/src/main/scala/example/ApiError.scala]: Done

**Review by mijicd — APPROVED** [review]

**Review by anovakovic01 — APPROVED** [review]

**Pull request merged**